### DISCUSSION TITLE
payments restructuring...

### DISCUSSION THREAD

**Discussion by proseLA** [zencart/zencart#5787] [Ideas]
after having coded a bunch of payment modules, i am curious whether people have thoughts to providing a structured framework, ie new database tables, for recording of payments and refunds.

i have seen it handled so many different ways, some of which are really not good.

that old super orders plugin had some good thoughts on it.

with the ability of most payment modules to issue partial refunds, and with store owners editing orders after receipt, i think it's very difficult to keep track of what has been paid.

in looking at the square payment module, there is a new table for keeping track of transaction IDs, but it does not even store the amount.  i think that's wrong.

storing details in the order comments table, i think is a bit of a hack; not the actual recording of the data in said table, but trying to figure out how much of an order total is paid v unpaid.

anyone else have thoughts on this?

**Comment by scottcwilson** [comment]
> after having coded a bunch of payment modules, i am curious whether people have thoughts to providing a structured framework, ie new database tables, for recording of payments and refunds.

Yes!  Please do so - it's helpful for tracking after the fact.

**Comment by brittainmark** [comment]
I would love to see this. Especially as I use square, and it does not record the refunds in zen cart.

**Comment by proseLA** [comment]
> I would love to see this. Especially as I use square, and it does not record the refunds in zen cart.

from what i looked at the square module only allows a void or a full refund.  there is no partial refund...  or am i wrong on that?

**Comment by robryanx** [comment]
A few structural ideas based on what I have done:
- A payments table which collects the common fields - for example I have `payments` and then say `payments_extra_stripe`
- A separate refunds table, so that multiple refunds per order can be supported - these can all be shown as part of the admin notification along with any refund errors.
- Storing the payment earlier in the process and storing the id as part of the session - say in `pre_confirmation_check` - so that errors errors at that step the `before_process` can both be stored and cases where someone has dropped out, useful for support.
- Multiple accounts - not sure how common this is for others but most of the payment providers we use we end up having multiple accounts to cover different entity and currency setups.
- Some standardisation with traits - some aspects of admin notification, such as some of the display and `_doRefund` method basically the same for every provider.
- A bit more complicated but there is some possibility of standardising tokenised stored cards - for example I use Stripe/PIN Payments and Airwallex which all have similar APIs around stored cards.

As an aside to this, it would probably simplify and improve paypal a lot to do an integration with the JS SDK and REST API, obviously a big undertaking though.

**Comment by proseLA** [comment]
hey rob,
thanks for the input!  a few things:

- this is really a schema idea for ZC, and as such it is really about having ZC specified table(s) for the storage of payment data.
- it is NOT an enforcement of process.  as you are no doubt aware, most payment modules are part of the ZC plugin infrastructure.  so when the payment gets stored is more about the module than what i am proposing.
- with regards to the tables, i'm in agreement, and think a separate refunds table is the way to go.
- i do not think the person/people who approve payment plugins will be enforcing the use of the tables...  although i could be wrong.... 
- storing of tokens is another possibility for enhancement.  despite having coded multiple payment modules that make use of tokens, i'm not inclined to do that at all.  i think that is more about the process of charging a token, and as such not really what my goal/thoughts are here.
- its really about having a common payments and refunds table.  that way when one instantiates an order object, all payments and refunds for that order (assuming they are all in the tables), are part of the order object, and anyone can know the order total, how much was paid, and how much is owed.
- right now, there is a lack of transparency to seeing all of that data due to the numerous ways payment modules handle it.
- i debating adding a few extra fields for unforeseen elements of various payment objects.
- i'm not sure what you are storing in `payments_extra_stripe`, but i hope it is not the magnetic card stripe data!

thanks for the input!

**Comment by robryanx** [comment]
Sorry to clarify a bit what I meant with `payments_extra_stripe`, I mean for for the Stripe payment gateway for credit cards.

So the `payments` table would be mostly common fields between most providers, such as status, amount, transaction id, details on the card (country/bin/exp etc).
And for Stripe specifically it has things like the specific fraud data fields that they provide.

That way you can standardise across some standard fields, but have a structure for plugins to extend that.

I understand what you are saying that you can at least have some of that, the payment amount itself and the refund amounts, without requiring any editing of existing modules which would probably make sense as a first step.

**Comment by brittainmark** [comment]
> 
With Square...
You can do a partial refund. You just set the amount and tick the box in details page of the order having shown additional payment details. I do it a lot if the postage is less than expected on my orders.

**Comment by proseLA** [comment]
> > 
> 
> With Square... You can do a partial refund. You just set the amount and tick the box in details page of the order having shown additional payment details. I do it a lot if the postage is less than expected on my orders.

really?  i will re-look at that.  i missed that what i was deep into last week.  i'll be back into that soon.
thanks for the info.

**Comment by proseLA** [comment]
ok @brittainmark, not sure how i missed it, but yes there is a partial refund.  thanks for pointing that out to me.

**Comment by proseLA** [comment]
i have given this some thought.  i was thinking about setting up an abstract class called `payments` or perhaps `zenPayments` that extends base.  within this abstract class, we can set up the 2 tables - payments and refunds, and create a few methods for storing and updating information in said table.  then any payment module that wants to make use of these tables can in turn extend the abstract class as opposed to `base`.

thoughts?

**Comment by scottcwilson** [comment]
I'd say just build the tables as required for Square, since any other payment method might have different requirements and would thus need its own tables.

**Comment by proseLA** [comment]
> I'd say just build the tables as required for Square, since any other payment method might have different requirements and would thus need its own tables.

yeah, not going to do that.  not really sure what that means.

if you look at @brittainmark s comment:

> I would love to see this. Especially as I use square, and it does not record the refunds in zen cart.

this is the essence of the problem.  the square table does not even record the amount charged/collected for the initial payment.  the only way to see anything is to make an api call to square and get all of the amounts.  which seems a waste of bandwidth to me.  which is why in my auth.net module i record everything in tables on zen-cart.

one always could get into the possibility of having different amounts recorded in ZC v the payment providers dashboard (which is always the correct amount), but any robust accounting system will always record numbers in their own books, and hopefully the developer does it correctly.

these 2 new tables are for storing of payments and refunds so that a store owner can see the true amount owed on an order.  in addition, developers can use them to store transaction ids in use for their processing...  or not.  can is the operative word.

they are NOT meant as a replacement for tables for any payment method....  although they can be.  that would be up to the developer.  ideally, they would be structured with enough fields to do that, but again it is the developer's call.

**Comment by brittainmark** [comment]
For me something that records the payment method, the amount, the order id would be a minimum. Ideally, the transaction id for the payment method would also be useful. Having it in one place for payment and a second for refunds would make my life a lot easier for me to reconcile the payments. I know it is possible to issue a refund directly from Square so would create an issue in accounting if this was done. But at the moment, it is very time-consuming to try to reconcile the square payments/refunds to the actual orders.

**Comment by robryanx** [comment]
> i have given this some thought. i was thinking about setting up an abstract class called `payments` or perhaps `zenPayments` that extends base. within this abstract class, we can set up the 2 tables - payments and refunds, and create a few methods for storing and updating information in said table. then any payment module that wants to make use of these tables can in turn extend the abstract class as opposed to `base`.
> 
> thoughts?

Or just as a trait. You could provide a standard set of payments behaviors that each payments class can opt in to. This could also flow into the admin notifications. As in a payments class uses your method to store refunds, maybe they can also just call a method so that those refunds are listed as part of the admin payment notification.
